File: atcoder/arc097/d.py
#!/usr/bin/env python3
# https://arc097.contest.atcoder.jp/tasks/arc097_b
import resource, sys
resource.setrlimit(resource.RLIMIT_STACK, [0x10000000, resource.RLIM_INFINITY])
sys.setrecursionlimit(100000)

n, m = map(int, input().split())
p = [int(x) - 1 for x in input().split()]
g = [list() for _ in range(n)]
for _ in range(m):
    u, v = map(int, input().split())
    u -= 1
    v -= 1
    g[u].append(v)
    g[v].append(u)
v = [False] * n
s = set()
def dfs(i):
    if v[i]: return
    s.add(i)
    v[i] = True
    for k in g[i]: dfs(k)

c = 0
for i in range(n):
    if v[i]: continue
    s = set()
    dfs(i)
    for k in s:
        if p[k] in s: c += 1
print(c)

File: atcoder/arc097/d.cc
// https://arc097.contest.atcoder.jp/tasks/arc097_b
#include <iostream>
#include <vector>

using namespace std;

typedef vector<int> vi;
typedef vector<vi> vvi;

vvi g;
vi v;

void dfs(int u, int k) {
	v[u] = k;
	for (auto z:g[u]) {
    if (v[z]) continue;
    dfs(z, k);
  }
}

int main() {
	int n, m;
	cin >> n >> m;
	vi a(n);
	for (int i = 0; i < n; i++) {
		cin >> a[i];
		a[i]--;
	}
	g = vvi(n);
	for (int i = 0; i < m; i++) {
		int u, v;
		cin >> u >> v;
		u--;
		v--;
		g[u].push_back(v);
		g[v].push_back(u);
	}
	v = vi(n);
  int k = 1;
	for (int i = 0; i < n; i++) {
		if (v[i]) continue;
		dfs(i, k);
    k++;
	}
	int c = 0;
  for (int i = 0; i < n; i++) c += v[i] == v[a[i]];
	cout << c << endl;
}
